src/build.ts

```diff
@@ -0,0 +1,105 @@
+const writeFileAsync = util.promisify(fs.writeFile)
+const chmodAsync = util.promisify(fs.chmod)
+const mkdirAsync = util.promisify(fs.mkdir)
+const execFile = util.promisify(childProcess.execFile)
+
+function transformFileAsync(filename: string, opts: babel.TransformOptions): Promise<babel.BabelFileResult | null> {
+  return new Promise((resolve, reject) => {
+    babel.transformFile(filename, opts, (err, result) => {
+      if (err) {
+        return reject(err)
+      }
+      resolve(result)
+    })
+  })
+}
+
+interface BuildEntry {
+  inFile: string
+  outFile: string
+  file: string
+  mode: number
+  state: string
+}
+
+export async function babelBuild(rootDirs: string[], outDir: string): Promise<BuildEntry[]> {
+  const buildList: BuildEntry[] = []
+  for (const rootDir of rootDirs) {
+    const fullOutDir = path.resolve(outDir, rootDir)
+    const fullRootDir = path.resolve(rootDir)
+    const dirs: string[] = [fullRootDir]
+    while (dirs.length > 0) {
+      const dir = dirs.shift() as string
+      for (const file of await readdirAsync(dir)) {
+        const inFile = path.resolve(dir, file)
+        const inStat = await statAsync(inFile)
+        if (inStat.isDirectory()) {
+          dirs.push(inFile)
+        } else if (inStat.isFile()) {
+          if (!file.match(/\.[tj]sx?$/) || file.match(/\.(?:test|d)\.[tj]s$/)) {
+            continue
+          }
+          const relative = path.relative(fullRootDir, inFile)
+          const outFile = path.join(fullOutDir, relative).replace(/ts(x?)/, 'js$1')
+          const entry: BuildEntry = {
+            inFile,
+            outFile,
+            file: `${path.join(rootDir, relative)}`,
+            mode: inStat.mode,
+            state: ''
+          }
+          try {
+            const outStat = await statAsync(outFile)
+            if (inStat.mtime > outStat.mtime) {
+              entry.state = 'updated'
+            } else {
+              entry.state = 'same'
+            }
+          } catch (e) {
+            entry.state = 'new'
+          }
+          buildList.push(entry)
+        }
+      }
+    }
+  }
+  for (const build of buildList) {
+    const result = await transformFileAsync(build.inFile, {
+      sourceFileName: path.join(
+        path.relative(path.dirname(build.outFile), path.dirname(build.inFile)),
+        path.basename(build.inFile)
+      ),
+      ignore: ['**/*.d.ts', 'src/**/*.test.ts'],
+      sourceMaps: true,
+      caller: {
+        name: '@babel/cli'
+      }
+    })
+    if (!result) {
+      // File ignored
+      continue
+    }
+    const mapFile = build.outFile + '.map'
+    result.code = result.code + `\n//# sourceMappingURL=${path.basename(mapFile)}`
+    if (result.map) {
+      result.map.file = path.basename(build.outFile)
+    }
+    await mkdirAsync(path.dirname(build.outFile), { recursive: true })
+    await writeFileAsync(mapFile, JSON.stringify(result.map))
+    await writeFileAsync(build.outFile, result.code)
```

Comment:
These two could be run in parallel.

Reply:
The file writing should have rather limited overhead as the main optimization is only building the changed files.

Reply:
I this case I would go with readability, but we could try to see if it made a difference.